Backported for 2.6.14 by Kerin Millar <kerframil@gmail.com>:

From stable-bounces@linux.kernel.org  Sat Jan 21 13:18:41 2006
Date: Sat, 21 Jan 2006 13:13:13 -0800
From: Andrew Morton <akpm@osdl.org>
To: stable@kernel.org
Message-Id: <20060121131313.551c624f.akpm@osdl.org>
Cc: Andi Kleen <ak@muc.de>
Subject: [PATCH] Mask off GFP flags before swiotlb_alloc_coherent

From: Andi Kleen <ak@muc.de>

Mask off GFP flags before swiotlb_alloc_coherent

Signed-off-by: Andi Kleen <ak@suse.de>
Signed-off-by: Chris Wright <chris@sous-sol.org>
---

akpm: Guys, this is a simple patch from Andi to fix a gfp.h BUG due to us
passing bad gfp flags into gfp_zone.

This is not a 2.6.16-x backport - because the pci-gart code in mainline gt
radically redone.

This is, I think, a new problem for -stable.  This code hasn't had the
normal external testing...

--- linux-2.6.14.orig/arch/x86_64/kernel/pci-gart.c	2005-10-27 19:02:08.000000000 -0500
+++ linux-2.6.14/arch/x86_64/kernel/pci-gart.c	2006-01-27 09:40:55.000000000 -0500
@@ -238,6 +238,7 @@ dma_alloc_coherent(struct device *dev, s
 					   get_order(size));
 
 				if (swiotlb) {
+					gfp &= ~GFP_DMA;
 					return
 					swiotlb_alloc_coherent(dev, size,
 							       dma_handle,
